### PR TITLE
Specialized classes for single column big int join

### PR DESCRIPTION
**Pull request by skrzypo987**
## Description

Majority of joins in TPC benchmarks (and probably overall) are joins over s single bigint column. This PR introduces a set of classes that are specialized for this case to provide performance gain

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement
> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core engine
> How would you describe this change to a non-technical end user or system administrator?

Increase the performance of join on single bigint column
## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Improve performance of joins over a single BIGINT column
```

### CODE REVIEW

**Comment by skrzypo987** [comment]
[Benchmark-bigint-join.pdf](https://github.com/trinodb/trino/files/9111696/Benchmark-bigint-join.pdf)
TPC benchmarks.
Ignore the tpch results for results that contain "tpcds"

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 1]
```diff
@@ -1036,14 +1056,21 @@ public LookupSourceSupplier createLookupSourceSupplier(
...
                 Optional<JoinFilterFunctionFactory> filterFunctionFactory,
                 Optional<Integer> sortChannel,
                 List<JoinFilterFunctionFactory> searchFunctionFactories,
                 HashArraySizeSupplier hashArraySizeSupplier)
         {
             PagesHashStrategy pagesHashStrategy = pagesHashStrategyFactory.createPagesHashStrategy(channels, hashChannel);
             try {
-                return constructor.newInstance(session, pagesHashStrategy, addresses, channels, filterFunctionFactory, sortChannel, searchFunctionFactories, hashArraySizeSupplier);
+                if (singleBigintColumn) {
+                    checkArgument(joinChannels.size() == 1);
```
Comment (radek-starburst) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1068]: Maybe we should make assertion for type too?

**Review by radek-starburst — COMMENTED** [review]
We copy some code in that PR but I think it is better and more readable to do it in comparison to adding a new abstraction or any other way to share the code. 
LGTM.

File: core/trino-main/src/main/java/io/trino/operator/PagesIndex.java [hunk 2]
```diff
@@ -521,6 +521,7 @@ public LookupSourceSupplier createLookupSourceSupplier(
                     valueAddresses,
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/PagesIndex.java:521]: nit: still, I'm curious about microbenchmarks :). I recall I did something similar once and it didn't help

File: core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHash.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+import javax.annotation.Nullable;
+
+import java.util.Optional;
+
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes arrays used in the hash are always a power of 2
+public final class BigintJoinHash
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHash.java:28]: Is it any different from `JoinHash`? If not, I think we can avoid `BigintJoinHash` because `JoinHash` class is isolated, see: `io.trino.sql.gen.JoinCompiler#internalCompileLookupSourceFactory`

File: core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHashSupplier.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+                    addresses.size(),
+                    pagesHashStrategy,
+                    addresses);
+        }
+        else {
+            positionLinksFactoryBuilder = ArrayPositionLinks.builder(addresses.size());
+        }
+
+        this.pages = channelsToPages(channels);
+        this.pagesHash = new BigintPagesHash(addresses, pagesHashStrategy, positionLinksFactoryBuilder, hashArraySizeSupplier, pages, joinChannel);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHashSupplier.java:75]: Can you just add one line specialization in `JoinHashSupplier`?

**Review by sopel39 — COMMENTED** [review]
mostly lgtm % comments

**Comment by skrzypo987** [comment]
Most comments addressed

File: core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java [hunk 5]
```diff
@@ -71,7 +73,12 @@ public JoinHashSupplier(
         }
 
         this.pages = channelsToPages(channels);
-        this.pagesHash = new PagesHash(addresses, pagesHashStrategy, positionLinksFactoryBuilder, hashArraySizeSupplier);
+        if (singleBigintJoinChannel.isPresent()) {
+            this.pagesHash = new BigintPagesHash(addresses, pagesHashStrategy, positionLinksFactoryBuilder, hashArraySizeSupplier, pages, singleBigintJoinChannel.getAsInt());
+        }
+        else {
+            this.pagesHash = new DefaultPagesHash(addresses, pagesHashStrategy, positionLinksFactoryBuilder, hashArraySizeSupplier);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java:80]: add javadoc to `DefaultPagesHash/MultiChannelPagesHash`

File: core/trino-main/src/main/java/io/trino/operator/PagesIndex.java (hunk 2 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/PagesIndex.java:521]: No clue. Micro and Macro benchmarks has long been independent. I'd guess that JIT is doing different magic tricks in different environments

File: core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHash.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
+/*
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/join/BigintJoinHash.java]: Made some changes in TestHashJoinOperator. PTAL

File: core/trino-main/src/main/java/io/trino/operator/join/DefaultPagesHash.java [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static io.trino.operator.SyntheticAddress.decodePosition;
+import static io.trino.operator.SyntheticAddress.decodeSliceIndex;
+import static io.trino.util.HashCollisionsEstimator.estimateNumberOfHashCollisions;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes arrays used in the hash are always a power of 2
+public final class DefaultPagesHash
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/join/DefaultPagesHash.java:39]: To be honest I think this name is misleading as it handles single channel as well. I'd much rather change `MultiChannelGroupByHash` to `DefaultGroupByHash`. WDYT?

**Comment by skrzypo987** [comment]
Comments addressed

File: core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    @Override
+    public double getExpectedHashCollisions()
+    {
+        return expectedHashCollisions;
+    }
+
+    @Override
+    public int getAddressIndex(int position, Page hashChannelsPage, long rawHash)
+    {
+        return getAddressIndex(position, hashChannelsPage);
```
Comment (lukasz-stec) [inline core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java:160]: doesn't it make sense to use `rawHash`?

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 9]
```diff
@@ -1019,12 +1021,14 @@ private BytecodeNode typeEqualsIgnoreNulls(
     {
         private final Constructor<? extends LookupSourceSupplier> constructor;
         private final PagesHashStrategyFactory pagesHashStrategyFactory;
+        private final boolean singleBigintColumn;
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1024]: remove. just make local decision in `createLookupSourceSupplier` below

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 10]
```diff
@@ -1036,14 +1040,21 @@ public LookupSourceSupplier createLookupSourceSupplier(
...
+                List<Integer> joinChannels,
                 Optional<JoinFilterFunctionFactory> filterFunctionFactory,
                 Optional<Integer> sortChannel,
                 List<JoinFilterFunctionFactory> searchFunctionFactories,
                 HashArraySizeSupplier hashArraySizeSupplier)
         {
             PagesHashStrategy pagesHashStrategy = pagesHashStrategyFactory.createPagesHashStrategy(channels, hashChannel);
             try {
-                return constructor.newInstance(session, pagesHashStrategy, addresses, channels, filterFunctionFactory, sortChannel, searchFunctionFactories, hashArraySizeSupplier);
+                if (singleBigintColumn) {
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1051]: you can check `joinChannels` here directly

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java [hunk 11]
```diff
@@ -189,38 +191,40 @@ public void testInnerJoin(boolean parallelBuild, boolean probeHashEnabled, boole
...
 
-    @Test
-    public void testInnerJoinWithRunLengthEncodedProbe()
+    @Test(dataProvider = "typesProvider")
+    public void testInnerJoinWithRunLengthEncodedProbe(Type type, IntFunction<Object> valueProvider, IntFunction<Object> rleValueProvider)
     {
         TaskContext taskContext = createTaskContext();
 
         // build factory
-        RowPagesBuilder buildPages = rowPagesBuilder(false, Ints.asList(0), ImmutableList.of(VARCHAR))
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:212]: This seems quite complicated. Let's just switch to BIGINT and force default/bigint pages hash explicitly. VARCHAR was chosen here for convenience. 
We can leave a single test or two with multi channel columns (I think they already exist).

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java [hunk 12]
```diff
@@ -349,29 +353,38 @@ public void testYield()
...
-                .row("b")
-                .row("c");
+                .row(a)
+                .row(b)
+                .row(c);
         BuildSideSetup buildSideSetup = setupBuildSide(nodePartitioningManager, parallelBuild, taskContext, buildPages, Optional.empty());
         JoinBridgeManager<PartitionedLookupSourceFactory> lookupSourceFactory = buildSideSetup.getLookupSourceFactoryManager();
 
         // probe factory
-        List<Type> probeTypes = ImmutableList.of(VARCHAR);
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:381]: ditto: let's swtich to bigint and force default/bigint supplier

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java (hunk 11 above)
Comment (skrzypo987) [reply core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:212]: Do you mean reverting changes in this class and just changing single-column tests to `BIGINT`?

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java (hunk 9 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1024]: There are no types available in `createLookupSourceSupplier` so no way of changing whether it is bigint

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java (hunk 10 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1051]: ^^

File: core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BigintPagesHash.class).instanceSize();
+    private static final DataSize CACHE_SIZE = DataSize.of(128, KILOBYTE);
+    private final LongArrayList addresses;
+    private final List<Block> joinChannelBlocks;
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int mask;
+    private final int[] key;
+    private final long[] values;
```
[the remark this answers is not in the document]
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java:51]: I made a prototype of this approach. However, the base was my local branch that already introduced batching, which changes the characteristics of memory access.
I made two versions - one with twice as big long array and the second one when two keys are packed into every third long value.
Both approaches showed regression in micro benchmarks, so I don't think this is worth doing.

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java (hunk 9 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:1024]: Can you pass `OptionalInt singleBigintJoinChannel` then here?

File: core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java [hunk 14]
```diff
@@ -0,0 +1,215 @@
...
+            PagesHashStrategy pagesHashStrategy,
+            PositionLinks.FactoryBuilder positionLinks,
+            HashArraySizeSupplier hashArraySizeSupplier,
+            List<Page> pages,
+            int joinChannel)
+    {
+        this.addresses = requireNonNull(addresses, "addresses is null");
+        this.pagesHashStrategy = requireNonNull(pagesHashStrategy, "pagesHashStrategy is null");
+        requireNonNull(pages, "pages is null");
+        ImmutableList.Builder<Block> joinChannelBlocksBuilder = ImmutableList.builder();
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java:68]: nit: use stream instead

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java [hunk 15]
```diff
@@ -189,24 +188,24 @@ public void testInnerJoin(boolean parallelBuild, boolean probeHashEnabled, boole
...
 
-    @Test
-    public void testInnerJoinWithRunLengthEncodedProbe()
+    @Test(dataProvider = "singleBigintLookupSourceProvider")
+    public void testInnerJoinWithRunLengthEncodedProbe(boolean singleBigintLookupSource)
     {
         TaskContext taskContext = createTaskContext();
 
         // build factory
-        RowPagesBuilder buildPages = rowPagesBuilder(false, Ints.asList(0), ImmutableList.of(VARCHAR))
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:212]: nit: this (`VARCHAR -> BIGINT`) should be separate commit

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java [hunk 16]
```diff
@@ -1314,6 +1313,34 @@ public static Object[][] testMemoryLimitProvider()
...
     }
 
+    @DataProvider(name = "singleBigintLookupSourceProvider")
+    public static Object[][] singleBigintLookupSourceProvider()
+    {
+        return new Object[][] {{true}, {false}};
+    }
+
+    @DataProvider(name = "hashJoinTestValuesAndsingleBigintLookupSourceProvider")
+    public static Object[][] hashJoinTestValuesAndsingleBigintLookupSourceProvider()
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:1323]: nit: consider writing cross product method for combining these. I'm pretty sure we have something like that in code

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java [hunk 17]
```diff
@@ -17,7 +17,6 @@
 import com.google.common.collect.ImmutableSet;
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:17]: do we have any multi channel tests left?

**Review by sopel39 — APPROVED** [review]
lgtm % comments.

Please improve PR descriptions, e.g:
```


    How would you describe this change to a non-technical end user or system administrator?

slow -> fast
```

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 18]
```diff
@@ -168,14 +179,17 @@ private LookupSourceSupplierFactory internalCompileLookupSourceFactory(List<Type
...
+                types.get(getOnlyElement(joinChannels)) == BIGINT;
         Class<? extends LookupSourceSupplier> joinHashSupplierClass = IsolatedClass.isolateClass(
                 new DynamicClassLoader(getClass().getClassLoader()),
                 LookupSourceSupplier.class,
                 JoinHashSupplier.class,
                 JoinHash.class,
-                DefaultPagesHash.class);
-
-        return new LookupSourceSupplierFactory(joinHashSupplierClass, new PagesHashStrategyFactory(pagesHashStrategyClass));
+                singleChannelBigint ? BigintPagesHash.class : DefaultPagesHash.class);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:190]: Do we need to isolate bigint hash? Does it call any interface with multiple implementations?

File: core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java (hunk 14 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java:68]: I did, but some reviewer wrote that:

> nit: stream might potentially be slow for small pages.

File: core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java (hunk 17 above)
Comment (skrzypo987) [reply core/trino-main/src/test/java/io/trino/operator/join/unspilled/TestHashJoinOperator.java:17]: Yes, there are 4 or 5

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java (hunk 18 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:190]: Isn't the fact that `PagesHash` is an interface relevant? TBH I don't know much about class isolation. It was isolated before so I just added the new class

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 19]
```diff
@@ -112,8 +116,15 @@
 
     @Inject
     public JoinCompiler(TypeOperators typeOperators)
+    {
+        this(typeOperators, true);
+    }
+
+    @VisibleForTesting
+    public JoinCompiler(TypeOperators typeOperators, boolean enableSingleChannelBigintLookupSource)
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:124]: package private

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java [hunk 20]
```diff
@@ -168,14 +179,17 @@ private LookupSourceSupplierFactory internalCompileLookupSourceFactory(List<Type
     {
         Class<? extends PagesHashStrategy> pagesHashStrategyClass = internalCompileHashStrategy(types, outputChannels, joinChannels, sortChannel);
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:181]: Maybe:
```
OptionalInt singleBigintJoinChannel = Optional.empty();
if (enableSingleChannelBigintLookupSource
    && joinChannels.size() == 1
    && types.get(getOnlyElement(joinChannels)) == BIGINT) {
  singleBigintJoinChannel = OptionalInt.of(getOnlyElement(joinChannels))
}
```
Then you don't need extra boolean where you need to figure out similar, but other variable name (`singleChannelBigint`)

**Review by sopel39 — APPROVED** [review]
lgtm % comment

File: core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java (hunk 19 above)
Comment (skrzypo987) [reply core/trino-main/src/main/java/io/trino/sql/gen/JoinCompiler.java:124]: It is invoked from a different package

**Comment by sopel39** [comment]
We should generalize condition to not only `BIGINTS` but other int types too (similar for aggr)

**Comment by skrzypo987** [comment]
> We should generalize condition to not only `BIGINTS` but other int types too (similar for aggr)

WDYM? Aggregation handles only bigints. I don't think anyone in the world would join on a double column. And for int we need a separate implementation which I believe is out of scope here

**Comment by sopel39** [comment]
> WDYM? Aggregation handles only bigints. I don't think anyone in the world would join on a double column. And for int we need a separate implementation which I believe is out of scope here

Why do you need separate implementation for `INT, TINYINT, SMALLINT`?
